x509/golang.go

```diff
@@ -0,0 +1,192 @@
+	// x509 library also encodes OU and CN in a different order if compared
+	// to OpenSSL, which is less of an issue, but still worth to adjust
+	// while we are at it.
+	cnBytes, err := asn1.MarshalWithParams(cn, "utf8")
+	subcommands.DieNotNil(err)
+	ouBytes, err := asn1.MarshalWithParams(ou, "utf8")
+	subcommands.DieNotNil(err)
+	var (
+		oidCommonName         = []int{2, 5, 4, 3}
+		oidOrganizationalUnit = []int{2, 5, 4, 11}
+	)
+	pkixAttrTypeValue := []pkix.AttributeTypeAndValue{
+		{
+			Type:  oidCommonName,
+			Value: asn1.RawValue{FullBytes: cnBytes},
+		},
+		{
+			Type:  oidOrganizationalUnit,
+			Value: asn1.RawValue{FullBytes: ouBytes},
+		},
+	}
+	return pkix.Name{ExtraNames: pkixAttrTypeValue}
+}
+
+func CreateFactoryCa(ou string) string {
+	priv := genAndSaveKey(FactoryCaKeyFile)
+	crtTemplate := x509.Certificate{
+		SerialNumber: genRandomSerialNumber(),
+		Subject:      marshalSubject("Factory-CA", ou),
+		NotBefore:    time.Now(),
+		NotAfter:     time.Now().AddDate(20, 0, 0),
+
+		BasicConstraintsValid: true,
+		IsCA:                  true,
+		KeyUsage:              x509.KeyUsageCertSign,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth, x509.ExtKeyUsageClientAuth},
+	}
+	factoryCaString := genCertificate(&crtTemplate, &crtTemplate, &priv.PublicKey, priv)
+	writeFile(FactoryCaCertFile, factoryCaString, 0400)
+	return factoryCaString
+}
+
+func CreateDeviceCa(cn string, ou string) string {
+	factoryKey := parsePemPrivateKey(readFile(FactoryCaKeyFile))
+	factoryCa := parsePemCertificate(readFile(FactoryCaCertFile))
+	priv := genAndSaveKey(DeviceCaKeyFile)
+	crtTemplate := x509.Certificate{
+		SerialNumber: genRandomSerialNumber(),
+		Subject:      marshalSubject(cn, ou),
+		Issuer:       factoryCa.Subject,
+		NotBefore:    time.Now(),
+		NotAfter:     time.Now().AddDate(10, 0, 0),
+
+		BasicConstraintsValid: true,
+		IsCA:                  true,
+		MaxPathLenZero:        true,
+		KeyUsage:              x509.KeyUsageCertSign,
+	}
+	crtPem := genCertificate(&crtTemplate, factoryCa, &priv.PublicKey, factoryKey)
+	writeFile(DeviceCaCertFile, crtPem, 0400)
+	return crtPem
+}
+
+func SignTlsCsr(csrPemBytes []byte) string {
+	csr := parsePemCertificateRequest(csrPemBytes)
+	factoryKey := parsePemPrivateKey(readFile(FactoryCaKeyFile))
+	factoryCa := parsePemCertificate(readFile(FactoryCaCertFile))
+	crtTemplate := x509.Certificate{
+		SerialNumber: genRandomSerialNumber(),
+		Subject:      csr.Subject,
+		Issuer:       factoryCa.Subject,
+		NotBefore:    time.Now(),
+		NotAfter:     time.Now().AddDate(10, 0, 0),
+
+		IsCA:        true,
+		KeyUsage:    x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment | x509.KeyUsageKeyAgreement,
+		ExtKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		DNSNames:    csr.DNSNames,
+	}
+	crtPem := genCertificate(&crtTemplate, factoryCa, csr.PublicKey, factoryKey)
+	return crtPem
+}
+
+func SignCaCsr(csrPemBytes []byte, csrFileName string) string {
```

Comment:
We need to decide if we _do everything via files_ versus _do everything via bytes_, and then _if we use bytes or strings everywhere_.
Each option has its pros and cons, so the decision would be tough.

I mean literally to do either of these two:
A. Use bytes:
```
func SignTlsCsr(csrPem, factoryKeyPem, factoryCaPem []byte) (crtPem []byte)
func SignCaCsr(csrPem, factoryKeyPem, factoryCaPem []byte) (crtPem []byte)
```
B. Use strings:
```
func SignTlsCsr(csrPem, factoryKeyPem, factoryCaPem string) (crtPem string)
func SignCaCsr(csrPem, factoryKeyPem, factoryCaPem string) (crtPem string)
```
C. Use files:
```
// factoryKeyFile and factoryCaFile can be constants, like now
func SignTlsCsr(csrPemFile string) (crtPemFile string)
func SignCaCsr(csrPemFile string) (crtPemFile string)
```

Just IMHO a situation were we mix those things leads to interface discrepancies.

That said, **_(!) I'm not strong on this (!!!)_**.
Except for the `SignCaCsr` thing, all other functions seem easy to understand.
So, we can move on "as is" and improve iteratively later.

@detsch just give it a thought, and if you don't see a fast way forward to address this comment - then it is better to ignore it. Think about the `x509/golang.go` interface; `x509/bash.go` is secondary.

Reply:
That inconsistency was bothering me too. Using our x509.* methods without relying on file names would be the best approach. However, we would have to use more temp files inside the `bash.go` methods, and the change would not apply to the `factory_ca.*` files, whose names are hardcoded inside the bash scripts. I'm tempted to leave the implementation as is to avoid messing too much with the bash-based execution flow.
I've changed all pem parameters from `[]byte` to `strings` though, no more bytes there.

Reply:
I expected an answer like that... which sounds reasonable to me.
I'm glad you were able to extract a "quick and valuable" bit from my comment. 💪

Reply:
I've approved your changes already.
So please, ask @doanac to take a second look, and we're fine to merge.